substrate/frame/revive/src/precompiles/custom/xcm.rs

```diff
@@ -0,0 +1,113 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::Origin;
+use crate::{
+	precompiles::{AddressMatcher, Error, Ext, ExtWithInfo, Precompile},
+	Config,
+};
+use alloc::vec::Vec;
+use alloy_core::sol;
+use codec::{DecodeAll, Encode};
+use core::{marker::PhantomData, num::NonZero};
+use xcm_builder::{ExecuteController, ExecuteControllerWeightInfo, SendController};
+pub use IXcm::IXcmCalls;
+
+pub struct XcmPrecompile<T>(PhantomData<T>);
+
+sol! {
+	/// @title Defines all functions that can be used to interact with XCM
+	/// @author Tiago Bandeira
+	/// @dev Parameters MUST use SCALE codec serialisation
+	interface IXcm {
+		/// @notice Execute an XCM message locally with the caller's origin
+		/// @param message The XCM message to send
+		function xcmExecute(bytes calldata message) external;
+
+		/// @notice Send an XCM message to a destination chain
+		/// @param destination The destination location, encoded according to the XCM format
+		/// @param message The XCM message to send
+		function xcmSend(bytes calldata destination, bytes calldata message) external;
+	}
+}
+
+impl<T: Config> Precompile for XcmPrecompile<T> {
+	type T = T;
+	const MATCHER: AddressMatcher = AddressMatcher::Fixed(NonZero::new(10).unwrap());
+	const HAS_CONTRACT_INFO: bool = false;
+	type Interface = IXcm::IXcmCalls;
+
+	fn call(
+		_address: &[u8; 20],
+		input: &Self::Interface,
+		env: &mut impl Ext<T = Self::T>,
+	) -> Result<Vec<u8>, Error> {
+		let origin = env.origin();
+		let xcm_origin = match origin {
+			Origin::Root => frame_system::RawOrigin::Root.into(),
+			Origin::Signed(account_id) => {
+				frame_system::RawOrigin::Signed(account_id.clone()).into()
+			},
+		};
+
+		match input {
+			IXcmCalls::xcmSend(IXcm::xcmSendCall { destination, message }) => {
+				let final_destination =
+					xcm::VersionedLocation::decode_all(&mut &destination[..])
+						.map_err(|_| Error::Revert("Invalid destination format".into()))?;
+
+				let final_message = xcm::VersionedXcm::<()>::decode_all(&mut &message[..])
+					.map_err(|_| Error::Revert("Invalid message format".into()))?;
+
+				<<T as Config>::Xcm>::send(
+					xcm_origin,
+					final_destination.into(),
+					final_message.into(),
+				)
+				.map(|message_id| message_id.encode())
+				.map_err(|_| {
+					Error::Revert(
+						"XCM send failed: destination or message format may be incompatible".into(),
+					)
+				})
+			},
+			IXcmCalls::xcmExecute(IXcm::xcmExecuteCall { message }) => {
+				let final_message = xcm::VersionedXcm::decode_all(&mut &message[..])
+					.map_err(|_| Error::Revert("Invalid message format".into()))?;
+
+				let weight_limit =
+					<<T as Config>::Xcm as ExecuteController<_, _>>::WeightInfo::execute();
```

Comment:
This is not right. This weight is static. We want the user to be able to specify the `max_weight` (or `weight_limit`) that their message will use. Like [moonbeam does it](https://docs.moonbeam.network/builders/interoperability/xcm/xcm-utils/)

Reply:
@franciscoaguirre what would the default proof size be?